### PR TITLE
[SPARK-31568][R][DOCS] Add detail about func/key in gapply to documentation

### PR DESCRIPTION
**Pull request by MichaelChirico**
### What changes were proposed in this pull request?

Improve documentation for `gapply` in `SparkR`

### Why are the changes needed?

Spent a long time this weekend trying to figure out just what exactly `key` is in `gapply`'s `func`. I had assumed it would be a _named_ list, but apparently not -- the examples are working because `schema` is applying the name and the names of the output `data.frame` don't matter.

As near as I can tell the description I've added is correct, namely, that `key` is an unnamed list.

### Does this PR introduce any user-facing change?

No? Not in code. Only documentation.

### How was this patch tested?

Not. Documentation only

### CODE REVIEW

File: R/pkg/R/DataFrame.R [hunk 1]
```diff
@@ -1669,29 +1667,43 @@ setMethod("dapplyCollect",
...
+#' The output of \code{func} must be a \code{data.frame} matching \code{schema} --
+#' in particular this means the names of the output \code{data.frame} are irrelevant
+#'
 #' @seealso \link{gapplyCollect}
 #' @examples
 #'
 #' \dontrun{
-#' Computes the arithmetic mean of the second column by grouping
-#' on the first and third columns. Output the grouping values and the average.
+#' # Computes the arithmetic mean of the second column by grouping
```
Comment (MichaelChirico) [inline R/pkg/R/DataFrame.R:1686]: Even though this code in `\dontrun{}` won't be evaluated by CRAN machines, still end users might want to copy-paste the code directly & expect it to be valid R code (this is the point of the `run.dontrun` argument for `example()`).

So I added the comment hash here in addition to the change describing `func` detailed above.

Comment (HyukjinKwon) [reply R/pkg/R/DataFrame.R:1686]: Yup, I think we should comment them out.

**Comment by HyukjinKwon** [comment]
ok to test

**Comment by HyukjinKwon** [comment]
Looks good, @MichaelChirico. Mind filing a [JIRA](https://issues.apache.org/jira/projects/SPARK/summary), and keeping the [Github PR format](https://github.com/apache/spark/blob/master/.github/PULL_REQUEST_TEMPLATE)? It's best to follow the standard, see also https://spark.apache.org/contributing.html.

File: R/pkg/R/DataFrame.R [hunk 2]
```diff
@@ -1700,8 +1712,8 @@ setMethod("dapplyCollect",
 #'     y <- data.frame(key, mean(x$b), stringsAsFactors = FALSE)
 #' }, schema)
 #'
-#' We can also group the data and afterwards call gapply on GroupedData.
-#' For Example:
+#' # We can also group the data and afterwards call gapply on GroupedData.
+#' # For Example:
```
Comment (HyukjinKwon) [inline R/pkg/R/DataFrame.R:1716]: nit but while we're here, For Example -> For example:

**Comment by MichaelChirico** [comment]
Thanks @HyukjinKwon . I had read the Contributing guide but the rules for a documentation-only request weren't as clear so I went for the approach used on Apache Airflow by instinct (it was the most recent Apache project I PR'd). I'll file a Jira shortly.

**Comment by HyukjinKwon** [comment]
Yeah, technically doc-only change can be fine with/without a JIRA. I thought you wanted to file a JIRA assuming from the title :-). It's usually best to file a JIRA anyway.

**Comment by MichaelChirico** [comment]
done & edited PR format

**Review by HyukjinKwon — DISMISSED** [review]
~~LGTM~~

File: R/pkg/R/DataFrame.R [hunk 3]
```diff
@@ -1669,29 +1667,43 @@ setMethod("dapplyCollect",
...
+#' \code{func} is a function of two arguments. The first, usually named \code{key}
+#' (though this is not enforced) corresponds to the grouping key, will be an
+#' unnamed \code{list} of \code{length(cols)} length-one objects corresponding
+#' to the grouping columns' values for the current group.
+#'
+#' The second, herein \code{x}, will be a local \code{\link{data.frame}} with the
+#' columns of the input not in \code{cols} for the rows corresponding to \code{key}.
+#'
+#' The output of \code{func} must be a \code{data.frame} matching \code{schema} --
+#' in particular this means the names of the output \code{data.frame} are irrelevant
```
Comment (HyukjinKwon) [inline R/pkg/R/DataFrame.R:1680]: Oh, sorry, can we fix `gapplyCollect` too while we're here?

Comment (HyukjinKwon) [reply R/pkg/R/DataFrame.R:1680]: Should be the same. You can reword the documentation there to point `gapply` out or just copy and paste for now

Comment (MichaelChirico) [reply R/pkg/R/DataFrame.R:1680]: I saw this but I wasn't sure what to do 😅

Comment (MichaelChirico) [reply R/pkg/R/DataFrame.R:1680]: Any reason not to merge the Rd for `gapplyCollect` to be in the same help as `gapply`?

Comment (HyukjinKwon) [reply R/pkg/R/DataFrame.R:1680]: Can be .. any good clean and consistent way is fine as long as we can clarify the differences - `gapplyCollect` is a short cut of `gapply` and `collect`; `gapplyCollect` can omit the return schema, and directly collects to the driver side. Except this, all are same.

Comment (HyukjinKwon) [reply R/pkg/R/DataFrame.R:1680]: We can just conservatively copy and paste the current doc changes too for now. I don't mind.

Comment (MichaelChirico) [reply R/pkg/R/DataFrame.R:1680]: OK. I think merging the Rd is the right way to go, and can increase awareness too -- actually I had never heard of `gapplyCollect` before filing this PR.

But I'll have to brush up on my `roxygen2` to get there from here, will follow up in a while.

Comment (HyukjinKwon) [reply R/pkg/R/DataFrame.R:1680]: Thanks!

File: R/pkg/R/functions.R [hunk 4]
```diff
@@ -3951,7 +3951,6 @@ setMethod("map_values",
 #' @rdname column_collection_functions
 #' @aliases map_zip_with map_zip_with,characterOrColumn,characterOrColumn,function-method
 #'
-#' @examples
```
Comment (MichaelChirico) [inline R/pkg/R/functions.R:3954]: this was throwing a warning from `roxygen2` about empty `@examples`.

should add an example, but just deleted for now

**Comment by MichaelChirico** [comment]
@HyukjinKwon finished merging the Rds. Checked the output of `roxygen2::roxygenize('.')` & looks OK. Thanks for pushing in a better direction.

**Review by HyukjinKwon — COMMENTED** [review]
@MichaelChirico, I just took a look, and it looks as below:

Screen Shot 2020-04-27 at 2 37 31 PM

Seems like we should use a better name than `gapply` to be consistent with other instances in this doc. After taking a close look, we might have to group `dapply` and `dapplyCollect` together as, for instance, "R native function executions".

We should probably make a group like here https://github.com/apache/spark/blob/f73b8a83481aae569b504fcddfff43a13af5b844/R/pkg/R/functions.R#L21-L35

How about we taking one step back, and just copy and paste the changes in `gapply` to `gapplyCollect` for now in this PR?
We can proceed about the grouping and deduplication in a separate PR.

**Comment by MichaelChirico** [comment]
really opened a can of worms here huh.

To me, it makes sense to leave the PR as is, then do the `dapply`/`dapplyCollect` centralization in another PR.

Have filed this follow-up Jira in any case:

https://issues.apache.org/jira/browse/SPARK-31568

I could also try combining them in this PR, I don't think it would be substantially difficult, but I don't know the rules for closing multiple Jiras in the same PR.

up to you.

**Comment by HyukjinKwon** [comment]
Yeah, let's just do it completely separately. Let's just go back to the original status (before 
51ee77b), and the just copy and paste the changes into `gapplyCollect`.

Thanks for creating a followup JIRA.

**Comment by MichaelChirico** [comment]
@HyukjinKwon done

**Review by HyukjinKwon — APPROVED** [review]
LGTM

**Comment by HyukjinKwon** [comment]
Merged to master, branch-3.0 and branch-2.4.

**Pull request closed without merging**